cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -117,6 +121,11 @@ public void put(String key, Value value) {
   }
 
   void put(String key, Value value, Scope scope) {
+    // Check if put operation is being performed inside the Spark executor
+    if (!putOperationAllowed) {
+      throw new UnsupportedOperationException("Put operation is not allowed from Spark executor.");
```

Comment:
It is specific to Spark in this case. For MapReduce we do it through `NonUpdatableToken` added as a part of the PR - #3339 

MapperWrapper and ReducerWrapper gets the BasicMapReduceContext from the MapReduceContextProvider. There we supply the NonUpdatableToken.